README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
drep
====

<!-- badges: start -->
[![Travis build status](https://travis-ci.org/mrc-ide/drep.svg?branch=master)](https://travis-ci.org/mrc-ide/drep) [![Codecov test coverage](https://codecov.io/gh/mrc-ide/drep/branch/master/graph/badge.svg)](https://codecov.io/gh/mrc-ide/drep?branch=master) <!-- badges: end -->

`drep` allows calculation of the reproduction number (*R*<sub>0</sub>) of the dengue virus from force of infection using the at-equilibrium number of primary, secondary, tertiary and quaternary infections in a population and their relative infectiousness. This is done assuming that dengue transmission is at equilibrium.

Installation
------------

You can install the development version of the `drep` package using `devtools`. First install `devtools`, if you don't already have it.

``` r
install.packages("devtools")
library(devtools)
```

Then, in a fresh R session, install the `drep` package.

``` r
devtools::install_github("mrc-ide/drep")
```

Background
----------

Accurate estimates of transmission intensity of dengue are needed to quantify the global burden of the disease and assess the impact of control strategies. Transmission intensity of an infectious disease is measured using the force of infection (the per capita rate at which susceptible individuals acquire infection, *λ*) or the basic reproduction number (the average number of secondary infections generated by one primary case entering a susceptible population, *R*<sub>0</sub>). *λ* and *R*<sub>0</sub> are related through the age structure of the human host population. While *λ* is computed by accounting for the demographic structure of the human population, *R*<sub>0</sub> is normalized. Both these measures can be used to estimate the global burden of dengue using a set of climatic and environmental predictors and a geostatistical model to produce a continuous surface of transmission intensity and calculate burden outputs (infections, symptomatic cases, hospitalizations). This package provides a set of functions to translate average force of infection (per serotype) into reproduction number. Future versions of the package will also allow calculation of burden estimates.

We make a number of assumptions:

1.  four dengue serotypes are in circulation
2.  there is a maxiumum number of four dengue infections per individual
3.  dengue transmission is at equilibrium
4.  force of infection is constant through time

Example
-------

This is a basic example which shows you how to to convert force of infection into *R*<sub>0</sub>. To do this we need:

-   the age structure of the human population
-   the relative infectiousness of the four dengue infections.

We assume that the population is divided up into 20 age groups and define the following parameters:

-   number of age groups in the country human population (`n_age_groups`)
-   the force of infection estimate (`FOI`)
-   the relative infectiousness of the four dengue infections (`phis`)

We asssume, as a starting point, that all four infections have the same infectiousness.

``` r
n_age_groups <- 20
pop <- 500000
FOI <- 0.0235
phis <- c(1, 1, 1, 1)
```

We then simulate some data, including

-   the lower and upper age limits (`l_lim` and `u_lim`) of the country age groups
-   the proportion of individuals in each age gorup (`f_j`)

``` r
l_lim <- seq(0, 95, length.out = n_age_groups)
u_lim <- seq(5, 100, length.out = n_age_groups)
n_j <- sample(1:50, n_age_groups, replace = TRUE)
f_j <- n_j / sum(n_j)
```

The *R*<sub>0</sub> can be calculated using the `calculate_R0` function. The `calculate_R0` function takes six arguments:

-   `FOI`: a numeric value of the force of infection estimate, which represents the average estimate per serotype
-   `f_j`: a numeric vector of the proportions of individuals in each age group
-   `l_lim`: a numeric vector of the lower limits of the country age groups
-   `u_lim`: a numeric vector of the upper limits of the country age groups
-   `phis`: a numeric vector of the relative infectiousness of the numbers of primary, secondary, tertiary and quaternary dengue infections.

``` r
R0 <- drep::calculate_R0(FOI, f_j, u_lim, l_lim, phis)
R0
#> [1] 2.761299
```


NEWS.md
`# drep 0.1.0

* Added a `NEWS.md` file to track changes to the package.


R/data.R
#' Geolocated force of infection estimates of dengue
#'
#' A dataset containing the values and other attributes of > 580
#' force of infection estimates of dengue. The variables are as follows:
#'
#' @format A data frame with 583 rows and 9 variables:
#' \describe{
#'    \item{data_id}{unique identifier}
#'    \item{type}{type of study where the force of infection estimate was sourced from (serology, caseReport, pseudoAbsence)}
#'    \item{longitude}{longitude of the location of the study}
#'    \item{latitude}{latitude of the location of the study}
#'    \item{ISO}{3-digits country code}
#'    \item{ID_0}{GADM administrative unit level 0 code}
#'    \item{ID_1}{GADM administrative unit level 1 code}
#'    \item{FOI}{force of infection estimate (0.00000000--0.06159744)}
#'    \item{population}{size of the human population in level 1 adm unit}
#'  }
#' @references \url{https://mrcdata.dide.ic.ac.uk/_dengue/dengue.php}
"foi"


#' Country age structures
#'
#' A dataset containing the proportions of individuals in each of twenty 5-year
#' age groups. The variables are as follows:
#'
#' @format A data with frame with 200 rows and 21 variables:
#' \describe{
#'    \item{ID_0}{GADM administrative unit level 0 code}
#'    \item{band_0_4}{proportion of individuals of age 0-4}
#'    \item{band_5_9}{proportion of individuals of age 5-9}
#'    \item{band_10_14}{proportion of individuals of age 10-14}
#'    \item{band_15_19}{proportion of individuals of age 15-19}
#'    \item{band_20_24}{proportion of individuals of age 20-24}
#'    \item{band_25_29}{proportion of individuals of age 25-20}
#'    \item{band_31_34}{proportion of individuals of age 31-34}
#'    \item{band_35_39}{proportion of individuals of age 35-39}
#'    \item{band_40_44}{proportion of individuals of age 40-44}
#'    \item{band_45_49}{proportion of individuals of age 45-49}
#'    \item{band_50_54}{proportion of individuals of age 50-54}
#'    \item{band_55_59}{proportion of individuals of age 55-59}
#'    \item{band_60_64}{proportion of individuals of age 60-64}
#'    \item{band_65_69}{proportion of individuals of age 65-69}
#'    \item{band_70_74}{proportion of individuals of age 70-74}
#'    \item{band_75_79}{proportion of individuals of age 75-79}
#'    \item{band_80_84}{proportion of individuals of age 80-84}
#'    \item{band_85_89}{proportion of individuals of age 85-89}
#'    \item{band_90_94}{proportion of individuals of age 90-94}
#'    \item{band_95_99}{proportion of individuals of age 95-99}
#' }
#' @references \url{https://population.un.org/wpp/Download/Standard/Population/}
"age_structure"


R/calculating_R0.R

# -----------------------------------------------------------------------------

# calculate_incidences

#' \code{calculate_incidences} calculates the per-capita incidence rates of primary,
#' secondary, tertiary and quaternary dengue infections in individuals of a
#' particular age group, given a force of infection.
#'
#' @title Calculate the incidence of primary, secondary, tertiary and quaternary
#'  dengue infections.
#'
#' @param FOI The force of infection value. Numeric.
#'
#' @param u_lim A numeric vector of the upper age limits of the population age
#'  groups.
#'
#' @param l_lim A numeric vector of the lower age limits of the population age
#'  groups.
#'
#' @return A list containing four different numeric vectors with the incidence
#'  of primary, secondary, tertiary and quaternary infections in each age group.
#'
#' @examples
#' #Assumes human population divided in 5-year age groups and 20 age groups
#'
#' age_groups <- 20
#' # a force of infection value
#' a_FOI_value <- 0.031
#' # the lower limits of the age groups
#' a <- seq(0, 95, length.out = age_groups)
#' # the upper limits of the age groups
#' b <- seq(5, 100, length.out = age_groups)
#' # run
#' calculate_incidences(a_FOI_value, a, b)
#'
#' @export


calculate_incidences <- function(FOI, u_lim, l_lim) {

  if (FOI <= 0) {
    stop("FOI must be greater than zero")

  }

  get_incidences <- function(infec_prob) {
    (infec_prob / 4) / (u_lim - l_lim)

  }

  p_I1 <- exp(-4 * FOI * l_lim) - exp(-4 * FOI * u_lim)

  p_I2 <- 4 * (exp(-3 * FOI * l_lim) - exp(-3 * FOI * u_lim)) -
    3 * (exp(-4 * FOI * l_lim) - exp(-4 * FOI * u_lim))

  p_I3 <- 6 * (exp(-2 * FOI * l_lim) - exp(-2 * FOI * u_lim)) +
    8 * (exp(-3 * FOI * u_lim) - exp(-3 * FOI * l_lim)) +
    3 * (exp(-4 * FOI * l_lim) - exp(-4 * FOI * u_lim))

  p_I4 <- 4 * (exp(-FOI * l_lim) - exp(-FOI * u_lim) +
                 exp(-3 * FOI * l_lim) - exp(-3 * FOI * u_lim)) +
    6 * (exp(-2 * FOI * u_lim) - exp(-2 * FOI * l_lim)) +
    (exp(-4 * FOI * u_lim) - exp(-4 * FOI * l_lim))

  all_probs <- list(p_I1, p_I2, p_I3, p_I4)

  lapply(all_probs, get_incidences)

}


# -----------------------------------------------------------------------------

# incidences_to_numbers

#' \code{incidences_to_numbers} calculates the \emph{numbers} of people from different
#' age groups experiencing their primary, secondary, tertiary or quaternary
#' dengue infection, from estimates of the incidence \emph{rates} of primary, secondary,
#' tertiary or quaternary infections in each age group.
#'
#' @title Convert incidences into numbers.
#'
#' @param incidences A numeric vector of the incidence rates of primary, secondary,
#' tertiary or quaternary dengue infections in each age group.
#'
#' @param n_j A numeric vector of the proportion of individuals in each age group
#' in the human population.
#'
#' @return Numeric.
#'
#' @export


incidences_to_numbers <- function(incidences, n_j) {

  incidences * n_j

}


# -----------------------------------------------------------------------------

# calculate_R0

#' \code{calculate_R0} calculates the dengue reproduction number using the
#' at-equilibrium number of primary, secondary, tertiary and quaternary infections
#' in a population and their relative infectiousness.
#'
#' @title Calculate the Dengue reproduction number (\eqn{R_0})
#'
#' @inheritParams calculate_incidences
#'
#' @inheritParams incidences_to_numbers
#'
#' @param phis A numeric vector of the relative infectiousness
#'   of primary, secondary, tertiary and quaternary dengue infections.
#'
#' @return Numeric.
#'
#' @export


calculate_R0 <- function(FOI, n_j, u_lim, l_lim, phis) {

  incids <- calculate_incidences(FOI, u_lim, l_lim)

  infection_numbers_j <- lapply(incids, incidences_to_numbers, n_j)

  total_infection_numbers <- vapply(infection_numbers_j, sum, numeric(1))

  N <- 1 # pop size

  FOI * N / (sum(total_infection_numbers * phis))

}


# -----------------------------------------------------------------------------

# calculate_infectiousness_sym_2x_asym

#' \code{calculate_infectiousness_sym_2x_asym} calculates the infectiousness values
#' of primary, secondary, tertiary and quaternary assuming that symptomatic
#' infections are twice as infecious as asymptomatic ones. The infectiousness of
#' secondary infectious equals 1.
#'
#' @title Calculate infectiousness when symptomatic infections are twice as
#'  infectious as asymptomatic ones.
#'
#' @param prop_sym_parms numeric vector of proportions of primary, secondary and
#' tertiary/quaternary infections which are symptomatic.
#'
#' @return Numeric.
#'
#' @export


calculate_infectiousness_sym_2x_asym <- function(prop_sym_parms) {

  w_1 <- prop_sym_parms[1]
  w_2 <- prop_sym_parms[2]
  w_3 <- prop_sym_parms[3]

  phi_2 <- 1
  phi_1 <- (w_1 * 2 + (1 - w_1)) / (w_2 * 2 + (1 - w_2))
  phi_3 <- (w_3 * 2 + (1 - w_3)) / (w_2 * 2 + (1 - w_2))

  c(phi_1, phi_2, phi_3, phi_3)

}


R/work_in_progress.R
calculate_infectiousness_wgts_for_sym_asym_assumption <- function(w_1, w_2, w_3){

  phi_2 <- 1
  phi_1 <- (w_1 * 2 + (1 - w_1)) / (w_2 * 2 + (1 - w_2))
  phi_3 <- phi_4 <- (w_3 * 2 + (1 - w_3)) / (w_2 * 2 + (1 - w_2))

  c(phi_1, phi_2, phi_3, phi_4)
}

get_age_band_bounds <- function(tags) {

  age_band_tags_split <- strsplit(tags, "[^0-9]+")

  age_band_tags_split_num <- lapply(age_band_tags_split, as.numeric)

  age_band_tags_split_num_mat <- do.call("rbind", age_band_tags_split_num)

  age_band_tags_split_num_mat[, 2:3]

}

wrapper_to_multi_factor_R0 <- function(
  x, foi_data, age_struct,
  age_band_tags, age_band_lower_bounds, age_band_upper_bounds, prob_fun){

  phi_1 <- x[1]
  phi_2 <- x[2]
  phi_3 <- x[3]
  phi_4 <- x[4]

  vec_phis <- c(phi_1, phi_2, phi_3, phi_4)

  n <- nrow(foi_data)

  vapply(seq_len(n),
         wrapper_to_R0,
         numeric(1),
         foi_data = All_FOI_estimates_3,
         age_struct = country_age_struc,
         age_band_lower_bounds = age_band_L_bounds,
         age_band_upper_bounds = age_band_U_bounds,
         age_band_tags = age_band_tgs,
         vec_phis = vec_phis,
         prob_fun = prob_fun)

}

wrapper_to_R0 <- function(
  i, foi_data, age_struct,
  age_band_tags, age_band_lower_bounds, age_band_upper_bounds,
  vec_phis, prob_fun){

  m_j <- age_struct[age_struct$ID_0 == foi_data[i, "ID_0"], age_band_tags]
  FOI <- foi_data[i, "FOI"]

  calculate_R0(
    FOI = FOI,
    N = 1,
    n_j = m_j,
    age_band_lower_bounds = age_band_lower_bounds,
    age_band_upper_bounds = age_band_upper_bounds,
    vec_phis = vec_phis,
    prob_fun = prob_fun)

}


R/calculating_burden.R

# -----------------------------------------------------------------------------

# calculate_infections

#' \code{calculate_infections} calculates the number of annual dengue infections
#' in a human population, given a force of infection.
#'
#' @title Calculate the number of dengue infections.
#'
#' @inheritParams calculate_R0
#'
#' @param N Size of human population. NUmeric.
#'
#' @return Numeric.
#'
#' @export


calculate_infections <- function(FOI,
                                 n_j,
                                 u_lim,
                                 l_lim,
                                 N){

  incids <- calculate_incidences(FOI, u_lim, l_lim)

  infection_numbers_j <- lapply(incids, incidences_to_numbers, n_j)

  total_infection_number <- vapply(infection_numbers_j, sum, numeric(1))

  sum(total_infection_number) * 4 * N

}


# -----------------------------------------------------------------------------

# calculate_cases

#' \code{calculate_cases} calculates the number of annual dengue symptomatic
#' infections (\emph{cases}) in a human population, given a force of infection.
#'
#' @title Calculate the number of dengue cases.
#'
#' @inheritParams calculate_R0
#'
#' @inheritParams calculate_infections
#'
#' @param weights_vec A numeric vector of the proportions of primary, secondary,
#' tertiary and quaternary infections which are symptomatic.
#'
#' @return Numeric.
#'
#' @export


calculate_cases <- function(FOI,
                            n_j,
                            u_lim,
                            l_lim,
                            weights_vec,
                            N){

  gamma_1 <- weights_vec[1]
  rho <- weights_vec[2]
  gamma_3 <- weights_vec[3]

  incids <- calculate_incidences(FOI, u_lim, l_lim)

  I1_rate <- incids[[1]]
  I2_rate <- incids[[2]]
  I3_rate <- incids[[3]]
  I4_rate <- incids[[4]]

  tot_incid_rate_j <- rho * I2_rate + (gamma_1 * I1_rate) + (gamma_3 * (I3_rate + I4_rate))

  case_number_j <- incidences_to_numbers(tot_incid_rate_j, n_j)

  sum(case_number_j) * 4 * N

}


# -----------------------------------------------------------------------------

# calculate_hosp_cases

#' \code{calculate_hosp_cases} calculates the number of annual dengue symptomatic
#' infections (\emph{cases}) requiring hospitalization in a human population,
#' given a force of infection.
#'
#' @title Calculate the number of dengue hospitalized cases.
#'
#' @inheritParams calculate_R0
#'
#' @inheritParams calculate_infections
#'
#' @inheritParams calculate_cases
#'
#' @param parms A numeric vector of the proportions of primary, secondary and
#' tertiary and quaternary infections requiring hospitalization.
#'
#' @return Numeric.
#'
#' @export


calculate_hosp_cases <- function(FOI,
                                 n_j,
                                 u_lim,
                                 l_lim,
                                 parms,
                                 weights_vec,
                                 N){

  gamma_1 <- weights_vec[1]
  rho <- weights_vec[2]
  gamma_3 <- weights_vec[3]

  Q_1 <- parms[1]
  Q_2 <- parms[2]
  Q_3 <- parms[3]

  incids <- calculate_incidences(FOI, u_lim, l_lim)

  I1_rate <- incids[[1]]
  I2_rate <- incids[[2]]
  I3_rate <- incids[[3]]
  I4_rate <- incids[[4]]

  tot_incid_rate_j <- Q_2 * rho * I2_rate + (Q_1 * gamma_1 * I1_rate) + (gamma_3 * Q_3 * (I3_rate + I4_rate))

  case_number_j <- incidences_to_numbers(tot_incid_rate_j, n_j)

  sum(case_number_j) * 4 * N

}


use.R

devtools::load_all()


# Define parameters -----------------------------------------------------------


l_lim <- seq(0, 95, 5)
u_lim <- seq(5, 100, 5)
phis <- c(1, 1, 1, 1)



# Simulate some data ----------------------------------------------------------


FOI <- 0.0235
age_groups <- 20
x <- sample(1:50, age_groups, replace = TRUE)
n_j <- x / sum(x)
pop <- 500000


# Run a test ------------------------------------------------------------------


R0 <- calculate_R0(FOI, pop, n_j, u_lim, l_lim, phis)

R0


# Now with real data ----------------------------------------------------------


FOI <- foi[4, "FOI"] # a point in Brazil

FOI_id0 <- foi[4, "ID_0"]

n_j <- age_structure[age_structure$ID_0 == FOI_id0, 2:ncol(age_structure)]

pop <- foi[4, "population"]

R0 <- calculate_R0(FOI, pop, n_j, u_lim, l_lim, phis)

R0


# -----------------------------------------------------------------------------

# now we assume that only primary and secondary infections are infectious and therefore contribute to transmission.
# This assumptions implies that if a tertiary or a secondary infections are not infectious.


phis <- c(1, 1, 0, 0)

R0 <- calculate_R0(FOI, pop, n_j, u_lim, l_lim, phis)

R0


# ~ R0 increases as the virus has only two attempts at producing a given force of infection.


# -----------------------------------------------------------------------------

#now we assume that all infections are infeectious but symptomatic infections are twice as infectious as asymptomatic ones.
# This requires to know the proportion of primary. secondary, tertiary and quaternary infectionsa which are asymptomatic.
# we source this value for these parameters from

prop_sym_parms <- c(0.45, 0.85, 0.15)

phis <- calculate_infectiousness_sym_2x_asym(prop_sym_parms)

R0 <- calculate_R0(FOI, pop, n_j, u_lim, l_lim, phis)

R0

# calculate cases -------------------------------------------------------------


table1 <- readRDS("R0_r_all_squares_2.rds")
table1[table1[, "cell"] == 300000,"1"]

table2 <- readRDS("C_num_all_squares_2.rds")
table2[table2[, "cell"] == 300000,"1"]

table3 <- readRDS("I_num_all_squares_2.rds")
table3[table3[, "cell"] == 300000,"1"]

FOI <- 0.01995189
n_j <- age_structure[age_structure$ID_0 == 185, 2:ncol(age_structure)]
pop <- 432


sym_to_asym_ratios <- list(gamma_1, rho, gamma_3, gamma_3)

incids <- calculate_incidences(FOI, u_lim, l_lim)

infs <- vapply(incids, incidences_to_numbers, numeric(1), n_j, pop)

R0 <- calculate_R0(FOI, pop, infs, phis)

R0

cases <- calculate_cases(incids, sym_to_asym_ratios, n_j, pop)

sum(infs)
sum(cases)

### add datasets
dat <- read.csv("All_FOI_estimates_linear_env_var_area.csv")
foi <- dat[,c(1:8)]
foi$population <- dat$population
devtools::use_data(foi, foi, overwrite = TRUE)

write.csv(foi, file.path("vignettes", "foi.csv"),
          row.names = FALSE) # for vignettes

dat2 <- read.csv("country_age_structure.csv")
age_structure <- dat2[setdiff(names(dat2), c("country", "Reference_date", "band_80_99", "band_85_99"))]
devtools::use_data(age_structure, age_structure, overwrite = TRUE)

write.csv(age_structure, file.path("vignettes", "age_structure.csv"),
          row.names = FALSE) # for vignettes

### create vignette

devtools::use_vignette("my-vignette")
devtools::build_vignettes()
